runtime/src/installed_scheduler_pool.rs

```diff
@@ -391,11 +486,72 @@ impl BankWithScheduler {
     }
 
     pub const fn no_scheduler_available() -> InstalledSchedulerRwLock {
-        RwLock::new(None)
+        RwLock::new(SchedulerStatus::Unavailable)
     }
 }
 
 impl BankWithSchedulerInner {
+    fn with_active_scheduler(
+        self: &Arc<Self>,
+        f: impl FnOnce(&InstalledSchedulerBox) -> ScheduleResult,
+    ) -> ScheduleResult {
+        let scheduler = self.scheduler.read().unwrap();
+        match &*scheduler {
+            SchedulerStatus::Active(scheduler) => {
+                // This is the fast path, needing single read-lock most of time.
+                f(scheduler)
+            }
+            SchedulerStatus::Stale(_pool, (result, _timings)) if result.is_err() => {
+                Err(SchedulerAborted)
+            }
+            SchedulerStatus::Stale(_pool, _result_with_timings) => {
+                drop(scheduler);
+
+                let context = SchedulingContext::new(self.bank.clone());
+                let mut scheduler = self.scheduler.write().unwrap();
+                scheduler.transition_from_stale_to_active(|scheduler_pool, result_with_timings| {
+                    scheduler_pool.register_timeout_listener(self.do_create_timeout_listener());
+                    scheduler_pool.take_resumed_scheduler(context, result_with_timings)
+                });
+                drop(scheduler);
+
+                self.with_active_scheduler(f)
```

Comment:
Prefer to not recurse when possible, although sometimes it is much simpler I think this code could be written straight-forwardly without it.

```rust
        let scheduler = self.scheduler.read().unwrap();
        match &*scheduler {
            SchedulerStatus::Active(scheduler) => f(scheduler),
            SchedulerStatus::Stale(_pool, (result, _timings)) if result.is_err() => {
                Err(SchedulerAborted)
            }
            SchedulerStatus::Stale(_pool, _result_with_timings) => {
                drop(scheduler);
                let context = SchedulingContext::new(self.bank.clone());
                let mut scheduler = self.scheduler.write().unwrap();
                scheduler.transition_from_stale_to_active(|scheduler_pool, result_with_timings| {
                    scheduler_pool.register_timeout_listener(self.do_create_timeout_listener());
                    scheduler_pool.take_resumed_scheduler(context, result_with_timings)
                });
                drop(scheduler);
                let scheduler = self.scheduler.read().unwrap();
                let SchedulerStatus::Active(active_scheduler) = &*scheduler else {
                    panic!("Scheduler transition failed");
                };
                f(active_scheduler)
            }
            SchedulerStatus::Unavailable => panic!("Scheduler Unavailable"),
        }
```

different calls to `f`, but to me this reads a lot cleaner? Unless there is some chance that between drop and read-lock re-acquire there is somehow a different transition?

Reply:
If that chance is there, the recursion seems like it could possibly lead to an indefinite loop.

Reply:
done: 2dbb57b692fa1e2177daf94ad914962796379abf